### PR TITLE
helm: Configuring hasKey to enable boolean Usage 

### LINKED ISSUE #28381: Impossible to configure cilium via helm values
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

It impossible to configure some cilium options like set `enable-k8s-networkpolicy` to `false` in cilium helm chart
As per documentation `enable-k8s-networkpolicy` is `true` by default, so in case one wants to disable it, nothing happens because helm template allows only to enable it.
[This](https://github.com/cilium/cilium/blob/v1.14.1/install/kubernetes/cilium/templates/cilium-configmap.yaml#L766-L768) allows only setting to `true` (which is `true` by default). 
I am pretty sure it is not the only option which might have similar issue.

### Cilium Version

1.14.x, 1.15.x

### Kernel Version

N/A

### Kubernetes Version

N/A

### Sysdump

not relevant

### Relevant log output

_No response_

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by hasan-alkama**
Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #28381 

```release-note
Fix bug where setting the `k8sNetworkPolicy` Helm value to false did not take effect
```

### CODE REVIEW

**Comment by hasan-alkama** [comment]
@gandro The PR regarding the issue assigned

**Comment by joestringer** [comment]
Hi @hasan-alkama , thanks for the PR!

A couple of style notes: For a PR title, it helps to communicate to reviewers "what does the PR change". Currently the title reflects the problem, which is usually better kept in the PR description. Perhaps something like "Convert Helm values from string to bool"? Also, for the release note, the text is put into the release note verbatim. Please remove the `<!--`, `-->` characters.

**Comment by joestringer** [comment]
I'm not exactly sure about the upgrade impact, but whenever we change types like this I think it is worth considering what, if anything, users must change when upgrading to the latest Helm chart with this change.

**Comment by joestringer** [comment]
@hasan-alkama thanks, your adjustments make sense. I'll leave it up to the codeowners for further review. You'll need to rebase and remove the merge commit before we merge this.

Just to let you know, Cilium is a fast-moving project, so at the bottom of the page the button "This PR is outdated" will almost always be showing up, and it will ask you to update the PR with the `main` branch. However, there's generally no need to keep rebasing the PR unless there is a specific file conflict. Each time you do an update, we'll need to re-trigger the tests which take some time to run, so in general updates should primarily be made just when you make a change to the commits, for instance following some feedback from reviewers.

**Review by gandro — APPROVED** [review]
Looks good to me, thanks! I've double checked that all values default to false, which means there should be no observable change in behavior

**Review by nathanjsweet — CHANGES_REQUESTED** [review]
Your commit title needs to be less than 75 characters. Otherwise this looks good.

**Comment by hasan-alkama** [comment]
Thanks @nathanjsweet  for pointing Out i have updated the commit title.

**Review by sayboras — APPROVED** [review]
Thanks

**Review by nathanjsweet — APPROVED** [review]

**Comment by hasan-alkama** [comment]
@joestringer Thanks , What could be the reason for CI failure? As i see some checks are not successful should we wait for CI Failure to solve before merging ?

**Comment by gandro** [comment]
> @joestringer Thanks , What could be the reason for CI failure? As i see some checks are not successful should we wait for CI Failure to solve before merging ?

Those are caused by spot instances being removed during tests. This is unrelated to your PR and there was a recently merged PR to fix that: https://github.com/cilium/cilium/pull/32553

Let me restart the tests using the trigger, hopefully that should fix the issue. If not, we might have to rebase on main (but please hold off on that for the moment as it would invalidate all other test results)

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml:818]: It seems the EKS CI failure is actually legitimate. `endpointRoutes` defaults to false, but `ipam:eni` (used by ci-eks) manually sets `enable-endpoint-routes:true` separately: https://github.com/cilium/cilium/blob/d9ea2182400c4262d5bed90001d5763924d29e3a/install/kubernetes/cilium/templates/cilium-configmap.yaml#L504

This means this change breaks EKS. Thinking about this some more, we probably only use the the `hasKey` pattern for booleans that default to true.

**Review by gandro — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-configmap.yaml
Comment (nathanjsweet) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:818]: Hmm, maybe we shouldn't merge this then. I'm all for simplifying code where it makes sense, but I'm worried that someone will copy this pattern blindly in the future and end up with the wrong result.

Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:818]: I think we should do it for at least the booleans that default to true, but not the ones that default to false.

The bug referenced in the original post is that it's literally impossible to disable KNP support via `k8sNetworkPolicy=false` and I think that's something we ought to fix. 

@hasan-alkama would you mind changing the PR to only cover `k8sNetworkPolicy` and drop the rest?

Comment (hasan-alkama) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:818]: @gandro Yes I was thinking the same, i will Update the PR based on changing only k8sNetworkPolicy and drop the rest.
Thanks for pointing out

Comment (gandro) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml:818]: Awesome, thank  you!

**Review by gandro — APPROVED** [review]

**Comment by gandro** [comment]
Removed the upgrade impact label and upgraded the release note description as the new version of the PR is much smaller in scope

**Review by gandro — DISMISSED** [review]
Oh, I just noticed - please squash your two commits into one

**Pull request merged**
